src/choice_map.cpp

```diff
@@ -0,0 +1,442 @@
+  return ExtractTree(tree_mask_ext);
+}
+
+Tree ChoiceMap::ExtractTree(ExpandedTreeMask &tree_mask_ext) const {
+  const auto dag_root_id = dag_.GetDAGRootNodeId();
+  Assert(tree_mask_ext.find(dag_root_id) != tree_mask_ext.end(),
+         "DAG Root Id does not exist in ExpandedTreeMask map.");
+  Assert(tree_mask_ext[dag_root_id][AdjacentNode::LeftChild] != NoId,
+         "DAG Root Id has no children in ExpandedTreeMask map.");
+  const auto dag_rootsplit_id = tree_mask_ext[dag_root_id][AdjacentNode::LeftChild];
+
+  std::unordered_map<size_t, bool> visited_left;
+  std::unordered_map<size_t, bool> visited_right;
+  std::unordered_map<size_t, Node::NodePtr> nodes;
+
+  for (const auto &[node_id, adj_node_ids] : tree_mask_ext) {
+    std::ignore = adj_node_ids;
+    visited_left[node_id] = false;
+    visited_right[node_id] = false;
+  }
+
+  size_t node_id_counter = dag_.TaxonCount();
+  auto next_node_id = NoId;
+  auto current_node_id = dag_root_id;
+  // Continue until left and right children of rootsplit node have been visited.
+  nodes[dag_rootsplit_id] = nullptr;
+  while (nodes[dag_rootsplit_id] == nullptr) {
+    // If right branch (and left branch) already visited, join child nodes and return
+    // up the tree.
+    if (visited_right[current_node_id]) {
+      const auto left_child_id =
+          tree_mask_ext[current_node_id][AdjacentNode::LeftChild];
+      const auto right_child_id =
+          tree_mask_ext[current_node_id][AdjacentNode::RightChild];
+      nodes[current_node_id] = Node::Join(nodes.at(left_child_id),
+                                          nodes.at(right_child_id), node_id_counter);
+      node_id_counter++;
+      next_node_id = tree_mask_ext[current_node_id][AdjacentNode::Parent];
+    }
+    // If left branch already visited, go down the right branch.
+    else if (visited_left[current_node_id]) {
+      visited_right[current_node_id] = true;
+      next_node_id = tree_mask_ext[current_node_id][AdjacentNode::RightChild];
+    }
+    // If node is a leaf, return up the the tree
+    else if (dag_.IsNodeLeaf(current_node_id)) {
+      nodes[current_node_id] = Node::Leaf(current_node_id);
+      next_node_id = tree_mask_ext[current_node_id][AdjacentNode::Parent];
+    }
+    // If neither left or right child has been visited, go down the left branch.
+    else {
+      visited_left[current_node_id] = true;
+      next_node_id = tree_mask_ext[current_node_id][AdjacentNode::LeftChild];
+    }
+    Assert(next_node_id != current_node_id, "Node cannot be adjacent to itself.");
+    current_node_id = next_node_id;
+  }
+
+  Node::NodePtr topology = nodes[dag_rootsplit_id];
+  topology->Polish();
+  Assert((nodes.size() + 1) == tree_mask_ext.size(),
+         "Invalid TreeMask-to-Tree: Topology did not span every node in "
+         "the TreeMask.");
+  Tree tree(topology, Tree::BranchLengthVector(nodes.size()));
+
+  return tree;
+}
+
+bool ChoiceMap::TreeIsValid(const Tree &tree, const bool is_quiet) const {
```

Comment:
Interesting to have this here... it's not about a choice map at all. I think that you are building a tree by specifying leaves and joining them... is it possible to make an invalid tree? If so we should think about the node API and perhaps some tests over there.

Reply:
It is possible to make an invalid tree by my constraints.  It needs to include all the leaf nodes in the DAG.  And it needs to strictly bifurcating (`NodePtr` topologies allows multi-furcations).  Node throw errors during construction through joins for the other types of invalid topologies, like edges that terminate at non-leaf nodes and duplicate leaves.

Reply:
As far as where it should go, you could have it in node, but you would also need to pass it a DAG to check against for validity.  Because of that, maybe SubsplitDAG is the best place for it?

Reply:
It seems to me that it only uses the taxon count from the DAG. Couldn't we just have that be an argument? 

If we put it in SubsplitDAG, I would have thought that it would be checking that the tree could be generated from the DAG, which this doesn't do. OTOH such a function could be helpful.

We should also note that this checks that a tree is bifurcating... it's "valid" to have a non-bifurcating tree though not in the context of the DAG.

Reply:
Right.  That is the only argument we need in the taxon count.

As far as testing whether ExtractTopology is doing the right thing, the ChoiceMap selection method is entirely naive, just selecting the "first" parent, left and right child for each node in the DAG.  The nature of this "first" neighbor is subject to the implementation of the DAG and even the ordering of the input, if generated from a Newick tree file.  

So it feels like if the topology is valid and could be generated from the DAG (I agree we should have that test), then we should accept the topology as correct.  I think when we have more deterministic ChoiceMap selection methods, we can have more concrete testing in this regard.

Reply:
OK, do we need topology validity testing if we check that it exists in the DAG? And are you good regarding how to test if something sits in the DAG or not?

Reply:
I think the test if it exists in the DAG should be sufficient without a Topology validation test. My thought for this "topology exists in DAG" test would be to check via PCSPs.  Check if all PCSPs in the topology are also in the DAG.  I think we would include verifying that topology reaches every leaf node in this test and eliminate the valid topology test.

Reply:
Great!